### PR TITLE
Optimise merge registers for migrations

### PR DESCRIPTION
**Pull request by janezpodhostnik**
I have changed the snapshot so that it uses a map of key to array index. Because of that when merging in changes we don't need to go through all entries of the array.

This doesn't seem to have an impact on snapshot creation, or on the case where a large portion of the snapshot has been changed. 

But for the case where a small portion of a large snapshot is changed, this is a ~44% improvement 

### Before
```
Benchmark_Merge/create_1000
Benchmark_Merge/create_1000-12         	      20	    233152 ns/op
Benchmark_Merge/create_100000
Benchmark_Merge/create_100000-12       	      20	  25301158 ns/op
Benchmark_Merge/create_10000000
Benchmark_Merge/create_10000000-12     	      20	4743582931 ns/op
Benchmark_Merge/merge_1000_10
Benchmark_Merge/merge_1000_10-12       	      20	     96302 ns/op
Benchmark_Merge/merge_1000_100
Benchmark_Merge/merge_1000_100-12      	      20	     79810 ns/op
Benchmark_Merge/merge_1000_1000
Benchmark_Merge/merge_1000_1000-12     	      20	    127331 ns/op
Benchmark_Merge/merge_100000_10
Benchmark_Merge/merge_100000_10-12     	      20	   6931900 ns/op
Benchmark_Merge/merge_100000_1000
Benchmark_Merge/merge_100000_1000-12   	      20	   7132483 ns/op
Benchmark_Merge/merge_100000_100000
Benchmark_Merge/merge_100000_100000-12 	      20	  13534448 ns/op
Benchmark_Merge/merge_10000000_10
Benchmark_Merge/merge_10000000_10-12   	      20	1326502848 ns/op
Benchmark_Merge/merge_10000000_10000
Benchmark_Merge/merge_10000000_10000-12         	      20	1263474750 ns/op
Benchmark_Merge/merge_10000000_10000000
Benchmark_Merge/merge_10000000_10000000-12      	      20	3662666250 ns/op
```

### After
```
Benchmark_Merge/create_1000
Benchmark_Merge/create_1000-12         	      20	    253258 ns/op
Benchmark_Merge/create_100000
Benchmark_Merge/create_100000-12       	      20	  30626292 ns/op
Benchmark_Merge/create_10000000
Benchmark_Merge/create_10000000-12     	      20	4893254794 ns/op
Benchmark_Merge/merge_1000_10
Benchmark_Merge/merge_1000_10-12       	      20	    125283 ns/op
Benchmark_Merge/merge_1000_100
Benchmark_Merge/merge_1000_100-12      	      20	     66300 ns/op
Benchmark_Merge/merge_1000_1000
Benchmark_Merge/merge_1000_1000-12     	      20	    120348 ns/op
Benchmark_Merge/merge_100000_10
Benchmark_Merge/merge_100000_10-12     	      20	   5339110 ns/op
Benchmark_Merge/merge_100000_1000
Benchmark_Merge/merge_100000_1000-12   	      20	   5314683 ns/op
Benchmark_Merge/merge_100000_100000
Benchmark_Merge/merge_100000_100000-12 	      20	  13250754 ns/op
Benchmark_Merge/merge_10000000_10
Benchmark_Merge/merge_10000000_10-12   	      20	 559134696 ns/op
Benchmark_Merge/merge_10000000_10000
Benchmark_Merge/merge_10000000_10000-12         	      20	 536417231 ns/op
Benchmark_Merge/merge_10000000_10000000
Benchmark_Merge/merge_10000000_10000000-12      	      20	2795973331 ns/op
```

### CODE REVIEW

**Review by fxamacker — COMMENTED** [review]
Thanks for optimizing!  I agree that this can speedup large snapshots having only a small portion changed.

Given this optimization has tradeoffs (uses more memory), can you include memory usage info in your benchmarks (`B/op` and `allocs/op` columns) and also add benchmark(s) for 100+ million payloads since largest account reached ~170 million payloads last summer?

I'm concerned Atree migration may have some performance regression for large accounts due to increased memory use.  I left some comments about this.

nit: `payload_shnapshot_test.go` file probably needs to be renamed.

File: cmd/util/ledger/util/payload_shnapshot_test.go [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+
+	create := func(b *testing.B, payloadsNum int) {
+		b.Run("create_"+strconv.Itoa(payloadsNum), func(b *testing.B) {
+			benchmarkCreate(b, payloadsNum)
+		})
+	}
+
+	create(b, 1000)
+	create(b, 100000)
+	create(b, 10000000)
```
[the remark this answers is not in the document]
Comment (janezpodhostnik) [reply cmd/util/ledger/util/payload_shnapshot_test.go:33]: I cant run that on my machine :), it starts using swap.

**Comment by janezpodhostnik** [comment]
@fxamacker I got some surpising results from the benchmarks
```
goos: darwin
goarch: arm64
pkg: github.com/onflow/flow-go/cmd/util/ledger/util
Benchmark_PayloadSnapshot
Benchmark_PayloadSnapshot/create_1000
Benchmark_PayloadSnapshot/create_1000/MapBasedPayloadSnapshot
Benchmark_PayloadSnapshot/create_1000/MapBasedPayloadSnapshot-12         	    6338	    191957 ns/op	  274523 B/op	    5995 allocs/op
Benchmark_PayloadSnapshot/create_1000/PayloadSnapshot
Benchmark_PayloadSnapshot/create_1000/PayloadSnapshot-12                 	    6096	    191220 ns/op	  266307 B/op	    5994 allocs/op
Benchmark_PayloadSnapshot/create_100000
Benchmark_PayloadSnapshot/create_100000/MapBasedPayloadSnapshot
Benchmark_PayloadSnapshot/create_100000/MapBasedPayloadSnapshot-12       	      58	  19676024 ns/op	24031418 B/op	  601641 allocs/op
Benchmark_PayloadSnapshot/create_100000/PayloadSnapshot
Benchmark_PayloadSnapshot/create_100000/PayloadSnapshot-12               	      54	  19810440 ns/op	23230660 B/op	  601646 allocs/op
Benchmark_PayloadSnapshot/create_10000000
Benchmark_PayloadSnapshot/create_10000000/MapBasedPayloadSnapshot
Benchmark_PayloadSnapshot/create_10000000/MapBasedPayloadSnapshot-12     	       1	3377343875 ns/op	2508686488 B/op	59999997 allocs/op
Benchmark_PayloadSnapshot/create_10000000/PayloadSnapshot
Benchmark_PayloadSnapshot/create_10000000/PayloadSnapshot-12             	       1	3325415833 ns/op	2428683392 B/op	59999996 allocs/op
Benchmark_PayloadSnapshot/merge_1000
Benchmark_PayloadSnapshot/merge_1000/changes_1000
Benchmark_PayloadSnapshot/merge_1000/changes_1000/MapBasedPayloadSnapshot
Benchmark_PayloadSnapshot/merge_1000/changes_1000/MapBasedPayloadSnapshot-12         	  287155	      4255 ns/op	   13160 B/op	      19 allocs/op
Benchmark_PayloadSnapshot/merge_1000/changes_1000/PayloadSnapshot
Benchmark_PayloadSnapshot/merge_1000/changes_1000/PayloadSnapshot-12                 	   68947	     17578 ns/op	   21168 B/op	      17 allocs/op
Benchmark_PayloadSnapshot/merge_1000/changes_1000#01
Benchmark_PayloadSnapshot/merge_1000/changes_1000#01/MapBasedPayloadSnapshot
Benchmark_PayloadSnapshot/merge_1000/changes_1000#01/MapBasedPayloadSnapshot-12      	   90618	     13487 ns/op	   20280 B/op	     135 allocs/op
Benchmark_PayloadSnapshot/merge_1000/changes_1000#01/PayloadSnapshot
Benchmark_PayloadSnapshot/merge_1000/changes_1000#01/PayloadSnapshot-12              	   45922	     26625 ns/op	   26656 B/op	     133 allocs/op
Benchmark_PayloadSnapshot/merge_1000/changes_1000#02
Benchmark_PayloadSnapshot/merge_1000/changes_1000#02/MapBasedPayloadSnapshot
Benchmark_PayloadSnapshot/merge_1000/changes_1000#02/MapBasedPayloadSnapshot-12      	   14796	     80913 ns/op	   81064 B/op	    1077 allocs/op
Benchmark_PayloadSnapshot/merge_1000/changes_1000#02/PayloadSnapshot
Benchmark_PayloadSnapshot/merge_1000/changes_1000#02/PayloadSnapshot-12              	   14058	     85434 ns/op	   71568 B/op	    1075 allocs/op
Benchmark_PayloadSnapshot/merge_100000
Benchmark_PayloadSnapshot/merge_100000/changes_100000
Benchmark_PayloadSnapshot/merge_100000/changes_100000/MapBasedPayloadSnapshot
Benchmark_PayloadSnapshot/merge_100000/changes_100000/MapBasedPayloadSnapshot-12     	   10000	    125977 ns/op	 1008488 B/op	      19 allocs/op
Benchmark_PayloadSnapshot/merge_100000/changes_100000/PayloadSnapshot
Benchmark_PayloadSnapshot/merge_100000/changes_100000/PayloadSnapshot-12             	     729	   1808184 ns/op	 1811120 B/op	      17 allocs/op
Benchmark_PayloadSnapshot/merge_100000/changes_100000#01
Benchmark_PayloadSnapshot/merge_100000/changes_100000#01/MapBasedPayloadSnapshot
Benchmark_PayloadSnapshot/merge_100000/changes_100000#01/MapBasedPayloadSnapshot-12  	    4810	    225082 ns/op	 1088072 B/op	    1339 allocs/op
Benchmark_PayloadSnapshot/merge_100000/changes_100000#01/PayloadSnapshot
Benchmark_PayloadSnapshot/merge_100000/changes_100000#01/PayloadSnapshot-12          	     724	   1669820 ns/op	 1874480 B/op	    1337 allocs/op
Benchmark_PayloadSnapshot/merge_100000/changes_100000#02
Benchmark_PayloadSnapshot/merge_100000/changes_100000#02/MapBasedPayloadSnapshot
Benchmark_PayloadSnapshot/merge_100000/changes_100000#02/MapBasedPayloadSnapshot-12  	     148	   8174259 ns/op	 7737897 B/op	  107287 allocs/op
Benchmark_PayloadSnapshot/merge_100000/changes_100000#02/PayloadSnapshot
Benchmark_PayloadSnapshot/merge_100000/changes_100000#02/PayloadSnapshot-12          	     100	  10168800 ns/op	 8221200 B/op	  107286 allocs/op
Benchmark_PayloadSnapshot/merge_10000000
Benchmark_PayloadSnapshot/merge_10000000/changes_10000000
Benchmark_PayloadSnapshot/merge_10000000/changes_10000000/MapBasedPayloadSnapshot
Benchmark_PayloadSnapshot/merge_10000000/changes_10000000/MapBasedPayloadSnapshot-12 	      37	  45845741 ns/op	100008808 B/op	      19 allocs/op
Benchmark_PayloadSnapshot/merge_10000000/changes_10000000/PayloadSnapshot
Benchmark_PayloadSnapshot/merge_10000000/changes_10000000/PayloadSnapshot-12         	       2	 531950084 ns/op	180011696 B/op	      17 allocs/op
Benchmark_PayloadSnapshot/merge_10000000/changes_10000000#01
Benchmark_PayloadSnapshot/merge_10000000/changes_10000000#01/MapBasedPayloadSnapshot
Benchmark_PayloadSnapshot/merge_10000000/changes_10000000#01/MapBasedPayloadSnapshot-12         	     100	  44647159 ns/op	100811753 B/op	   13337 allocs/op
Benchmark_PayloadSnapshot/merge_10000000/changes_10000000#01/PayloadSnapshot
Benchmark_PayloadSnapshot/merge_10000000/changes_10000000#01/PayloadSnapshot-12                 	       2	 624171354 ns/op	180650960 B/op	   13335 allocs/op
Benchmark_PayloadSnapshot/merge_10000000/changes_10000000#02
Benchmark_PayloadSnapshot/merge_10000000/changes_10000000#02/MapBasedPayloadSnapshot
Benchmark_PayloadSnapshot/merge_10000000/changes_10000000#02/MapBasedPayloadSnapshot-12         	       1	2533207125 ns/op	771151704 B/op	10728241 allocs/op
Benchmark_PayloadSnapshot/merge_10000000/changes_10000000#02/PayloadSnapshot
Benchmark_PayloadSnapshot/merge_10000000/changes_10000000#02/PayloadSnapshot-12                 	       1	3236515542 ns/op	819984192 B/op	10728240 allocs/op
PASS
```

They show that the reverse map based approach is equal to the original approach regarding allocations.

**Comment by fxamacker** [comment]
> @fxamacker I got some surpising results from the benchmarks

@janezpodhostnik  Amazing!  You mentioned a "~44% improvement" but these results from your benchmarks look awesome!  :rocket: :tada: 

```bash
/merge_10000000/changes_10000000/MapBasedPayloadSnapshot-12     37     45845741 ns/op    100008808 B/op    19 allocs/op
/merge_10000000/changes_10000000/PayloadSnapshot-12                         2    531950084 ns/op    180011696 B/op    17 allocs/op
```

I'll take a closer look after wrapping up some urgent items (I'm out next week).  Very curious to see maxRSS and total duration of migration using mainnet24 snapshot as input when run on gcp (with enough RAM).

**Comment by janezpodhostnik** [comment]
There was an error in the benchmark naming, so here are the results again:
```
goos: darwin
goarch: arm64
pkg: github.com/onflow/flow-go/cmd/util/ledger/util
Benchmark_PayloadSnapshot
Benchmark_PayloadSnapshot/create_1000
Benchmark_PayloadSnapshot/create_1000/MapBasedPayloadSnapshot
Benchmark_PayloadSnapshot/create_1000/MapBasedPayloadSnapshot-12         	    6345	    189806 ns/op	  274522 B/op	    5995 allocs/op
Benchmark_PayloadSnapshot/create_1000/PayloadSnapshot
Benchmark_PayloadSnapshot/create_1000/PayloadSnapshot-12                 	    6342	    188993 ns/op	  266306 B/op	    5994 allocs/op
Benchmark_PayloadSnapshot/create_100000
Benchmark_PayloadSnapshot/create_100000/MapBasedPayloadSnapshot
Benchmark_PayloadSnapshot/create_100000/MapBasedPayloadSnapshot-12       	      60	  19287565 ns/op	24035608 B/op	  601653 allocs/op
Benchmark_PayloadSnapshot/create_100000/PayloadSnapshot
Benchmark_PayloadSnapshot/create_100000/PayloadSnapshot-12               	      63	  19271337 ns/op	23229734 B/op	  601643 allocs/op
Benchmark_PayloadSnapshot/create_10000000
Benchmark_PayloadSnapshot/create_10000000/MapBasedPayloadSnapshot
Benchmark_PayloadSnapshot/create_10000000/MapBasedPayloadSnapshot-12     	       1	3756713333 ns/op	2508686488 B/op	59999997 allocs/op
Benchmark_PayloadSnapshot/create_10000000/PayloadSnapshot
Benchmark_PayloadSnapshot/create_10000000/PayloadSnapshot-12             	       1	3586450959 ns/op	2428683392 B/op	59999996 allocs/op
Benchmark_PayloadSnapshot/merge_1000
Benchmark_PayloadSnapshot/merge_1000/changes_10
Benchmark_PayloadSnapshot/merge_1000/changes_10/MapBasedPayloadSnapshot
Benchmark_PayloadSnapshot/merge_1000/changes_10/MapBasedPayloadSnapshot-12         	  277328	      4059 ns/op	   13160 B/op	      19 allocs/op
Benchmark_PayloadSnapshot/merge_1000/changes_10/PayloadSnapshot
Benchmark_PayloadSnapshot/merge_1000/changes_10/PayloadSnapshot-12                 	   69555	     17034 ns/op	   21168 B/op	      17 allocs/op
Benchmark_PayloadSnapshot/merge_1000/changes_100
Benchmark_PayloadSnapshot/merge_1000/changes_100/MapBasedPayloadSnapshot
Benchmark_PayloadSnapshot/merge_1000/changes_100/MapBasedPayloadSnapshot-12        	   92430	     13262 ns/op	   20424 B/op	     137 allocs/op
Benchmark_PayloadSnapshot/merge_1000/changes_100/PayloadSnapshot
Benchmark_PayloadSnapshot/merge_1000/changes_100/PayloadSnapshot-12                	   45949	     26313 ns/op	   26800 B/op	     135 allocs/op
Benchmark_PayloadSnapshot/merge_1000/changes_1000
Benchmark_PayloadSnapshot/merge_1000/changes_1000/MapBasedPayloadSnapshot
Benchmark_PayloadSnapshot/merge_1000/changes_1000/MapBasedPayloadSnapshot-12       	   15498	     78274 ns/op	   79992 B/op	    1065 allocs/op
Benchmark_PayloadSnapshot/merge_1000/changes_1000/PayloadSnapshot
Benchmark_PayloadSnapshot/merge_1000/changes_1000/PayloadSnapshot-12               	   13675	     86213 ns/op	   71008 B/op	    1063 allocs/op
Benchmark_PayloadSnapshot/merge_100000
Benchmark_PayloadSnapshot/merge_100000/changes_10
Benchmark_PayloadSnapshot/merge_100000/changes_10/MapBasedPayloadSnapshot
Benchmark_PayloadSnapshot/merge_100000/changes_10/MapBasedPayloadSnapshot-12       	   10000	    129852 ns/op	 1008488 B/op	      19 allocs/op
Benchmark_PayloadSnapshot/merge_100000/changes_10/PayloadSnapshot
Benchmark_PayloadSnapshot/merge_100000/changes_10/PayloadSnapshot-12               	     711	   1557916 ns/op	 1811120 B/op	      17 allocs/op
Benchmark_PayloadSnapshot/merge_100000/changes_1000
Benchmark_PayloadSnapshot/merge_100000/changes_1000/MapBasedPayloadSnapshot
Benchmark_PayloadSnapshot/merge_100000/changes_1000/MapBasedPayloadSnapshot-12     	    5739	    211038 ns/op	 1087880 B/op	    1335 allocs/op
Benchmark_PayloadSnapshot/merge_100000/changes_1000/PayloadSnapshot
Benchmark_PayloadSnapshot/merge_100000/changes_1000/PayloadSnapshot-12             	     720	   1962199 ns/op	 1874288 B/op	    1333 allocs/op
Benchmark_PayloadSnapshot/merge_100000/changes_100000
Benchmark_PayloadSnapshot/merge_100000/changes_100000/MapBasedPayloadSnapshot
Benchmark_PayloadSnapshot/merge_100000/changes_100000/MapBasedPayloadSnapshot-12   	     135	   8045177 ns/op	 7748296 B/op	  107503 allocs/op
Benchmark_PayloadSnapshot/merge_100000/changes_100000/PayloadSnapshot
Benchmark_PayloadSnapshot/merge_100000/changes_100000/PayloadSnapshot-12           	     100	  10226462 ns/op	 8231600 B/op	  107502 allocs/op
Benchmark_PayloadSnapshot/merge_10000000
Benchmark_PayloadSnapshot/merge_10000000/changes_10
Benchmark_PayloadSnapshot/merge_10000000/changes_10/MapBasedPayloadSnapshot
Benchmark_PayloadSnapshot/merge_10000000/changes_10/MapBasedPayloadSnapshot-12     	      50	  42728700 ns/op	100008808 B/op	      19 allocs/op
Benchmark_PayloadSnapshot/merge_10000000/changes_10/PayloadSnapshot
Benchmark_PayloadSnapshot/merge_10000000/changes_10/PayloadSnapshot-12             	       2	 627278292 ns/op	180011696 B/op	      17 allocs/op
Benchmark_PayloadSnapshot/merge_10000000/changes_10000
Benchmark_PayloadSnapshot/merge_10000000/changes_10000/MapBasedPayloadSnapshot
Benchmark_PayloadSnapshot/merge_10000000/changes_10000/MapBasedPayloadSnapshot-12  	     100	  44014605 ns/op	100811464 B/op	   13331 allocs/op
Benchmark_PayloadSnapshot/merge_10000000/changes_10000/PayloadSnapshot
Benchmark_PayloadSnapshot/merge_10000000/changes_10000/PayloadSnapshot-12          	       2	 582249084 ns/op	180650672 B/op	   13329 allocs/op
Benchmark_PayloadSnapshot/merge_10000000/changes_10000000
Benchmark_PayloadSnapshot/merge_10000000/changes_10000000/MapBasedPayloadSnapshot
Benchmark_PayloadSnapshot/merge_10000000/changes_10000000/MapBasedPayloadSnapshot-12         	       1	3392543333 ns/op	771181704 B/op	10728525 allocs/op
Benchmark_PayloadSnapshot/merge_10000000/changes_10000000/PayloadSnapshot
Benchmark_PayloadSnapshot/merge_10000000/changes_10000000/PayloadSnapshot-12                 	       1	3708955459 ns/op	819997808 B/op	10728524 allocs/op
PASS
```

**Comment by bluesign** [comment]
@janezpodhostnik I think also need to benchmark `get`

**Review by turbolent — APPROVED** [review]
Great work!

**Comment by janezpodhostnik** [comment]
Before I merge this I want to see how do we feel about only having the `MapBasedPayloadSnapshot` instead of having 2 implementations, since it look like there are no allocation penalties.

**Review by fxamacker — APPROVED** [review]
@janezpodhostnik  Nice! :+1:   Thanks for updating PR!

> Before I merge this I want to see how do we feel about only having the MapBasedPayloadSnapshot instead of having 2 implementations, since it look like there are no allocation penalties.

I'm in favor of it (based on your amazing benchmark results) :rocket: :+1:  It looks like this PR can speedup both atree migration and Cadence 1.0 migration with far less memory overhead than initially thought (for the payload counts benchmarked).

One remaining concern is that we should confirm our expectations by running a test migration of all payloads from all accounts using mainnet snapshot/checkpoint as input.

If you want, we can merge this PR and I can try migration on large gcp vm when I return.

File: cmd/util/ledger/util/util.go [hunk 2]
```diff
@@ -69,38 +68,6 @@ func (a *AccountsAtreeLedger) AllocateStorageIndex(owner []byte) (atree.StorageI
 	return v, nil
 }
 
-type PayloadSnapshot struct {
```
Comment (turbolent) [inline cmd/util/ledger/util/util.go:72]: 👌

**Comment by fxamacker** [comment]
@janezpodhostnik @turbolent 

> One remaining concern is that we should confirm our expectations by running a test migration of all payloads from all accounts using mainnet snapshot/checkpoint as input.

:warning: Unfortunately, using this optimization made atree migration ~20 minutes slower and use ~16 GB more RAM.

```
Before: MaxRSS is 2.07 TB, duration (including ~30m validation) is 4h23 
After:    MaxRSS is 2.09 TB, duration (including ~30m validation) is 4h43
```

I used Atree inlining + Cadence 0.42 using mainnet data (Jan 19, 2024 checkpoint) as input.

Test vm was large enough RAM=3.84 TB and I used nworkers=40 (for consistency with old benchmarks).

**Comment by janezpodhostnik** [comment]
Thank you for the info.

I'm not quite sure why this is. The benchmarks looked promising...

I will prepare a PR to revert this.

**Comment by turbolent** [comment]
Though it may have slight adverse effects on one migration (atree register inlining), maybe it's still helpful for others (e.g. updating contracts in the Cadence 1.0 migration)?

**Comment by fxamacker** [comment]
>  I'm not quite sure why this is. The benchmarks looked promising...
>  
>  I will prepare a PR to revert this.

@janezpodhostnik  Maybe revert only for atree migration because I didn't test this optimization with Cadence 1.0 migration.

It is probably best to test with mainnet data as input for Cadence 1.0 migrations.

**Comment by fxamacker** [comment]
> Though it may have slight adverse effects on one migration (atree register inlining), maybe it's still helpful for others (e.g. updating contracts in the Cadence 1.0 migration)?

@turbolent I agree and I also think we should test with mainnet data as input for Cadence 1.0 migrations to determine impact of this optimization.

Both approaches have tradeoffs and we may benefit by using the most appropriate approach for each migration or account size.

**Comment by fxamacker** [comment]
I'm going to try using account size at runtime to choose merge implementation.

I'll let you know how it goes!

**Pull request merged**